### PR TITLE
Wrappers

### PR DESCRIPTION
**Pull request by willrogers**
@TimGuiteDiamond a little rearrangement.

TooltipWrapper is now always added if a `tooltip` property is there.

I plan to remove the `wrappers` nested object, always add menu button and make alarm border a top-level boolean.

### CODE REVIEW

**Comment by TimGuiteDiamond** [comment]
In principle I am happy with these changes, but they would almost certainly have a performance impact so I'm not sure how much re-testing would be necessary.

It would be disingenuous to say that we can update this many PVs at this rate when those numbers were generated using a readback widget without tooltip or otherwise.

I guess at the least I should investigate the impact before we merge this branch?

**Comment by TimGuiteDiamond** [comment]
With respect to the nested objects, I am happy with unwrapping the wrappers etc but I think keeping containerStyling and widgetStyling separated and enclosed is useful to a widget writer because they can just inject `props.style` without too many worries. Otherwise we are going to need to extract quite a few props in each component and assign them again, increasing the chance that something gets missed and that things don't work properly if we change the internals of widget/pvwidget to provide a different set of style props. Something to think about?

**Comment by willrogers** [comment]
Disingenuous eh? Surely you were already testing without a tooltip.

I think the point is that we need to make sure these performance tests are repeatable without too much grief. It will always be useful to be able to benchmark any changes we make.

Re styles: I'm not exactly sure of the pros and cons here but explicit properties are nice because you know exactly what is supported.

**Comment by TimGuiteDiamond** [comment]
Yes the current tests don't use a tooltip, I would imagine that will slow it down a bit, maybe not a lot? It would be quite important that it works really well if we make it the default and I just don't have the numbers right now, but could investigate so we can make a decision before merging?

I implemented flat Readback and flat Display along with flat widget and flat PV widget here: #3d7c86bb1 
I thought it just added a lot of work to building a component and therefore increases the chances that we will miss something. It is more explicit, I agree, but a lot of the props such as height, width, border size are not really to do with the component behaviour.

**Comment by willrogers** [comment]
I don't think it will slow things any more than other changes we are likely to make to the widget components.

So making it easy to benchmark is the key I think.

**Comment by TimGuiteDiamond** [comment]
Library for preparing benchmarking is here: https://gitlab.diamond.ac.uk/controls/python3/wdm_performance_test

File: src/components/TooltipWrapper/tooltipWrapper.tsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  if (props.resolvedTooltip) {
+    return (
+      <div
+        style={{
+          position: "relative",
+          height: "100%",
+          width: "100%",
+          ...style
+        }}
+      >
```
[the remark this answers is not in the document]
Comment (TimGuiteDiamond) [reply src/components/TooltipWrapper/tooltipWrapper.tsx]: See #113

**Comment by TimGuiteDiamond** [comment]
# Performance Testing

Checked render times against the readback with no wrappers and found only small differences:

| No. PVs | Tooltip (ms) | No Wrappers (ms) |
| --- | --- | --- |
| 1 | 7 | 8 |
| 10 | 22 | 23 |
| 20 | 35 | 25 |
| 100 | 269 | 211 |
| 200 | 546 | 525 |

This does not concern me that much

**Comment by TimGuiteDiamond** [comment]
However, when rendering with the tooltip open, this appears: 

CascadingUpdate

This is trying to tell us that we are triggering more updates within an update, and is generally a waste and a bad thing.

See these issues:
- https://github.com/reduxjs/react-redux/issues/834
- https://github.com/downshift-js/downshift/issues/686

This is worthy of investigation as it adds significant overhead to the render time when the tooltip is being displayed - one readback takes around 16ms when this is happening. This appears to be an approximately constant overhead, adding more widgets does not increase the time spent in these cascaded changes. Nevertheless it is indicative of the fact that we could write this part of the code better.

**Comment by TimGuiteDiamond** [comment]
This issue arises from the [componentDidUpdate lifetime hook in the Popover package](https://github.com/alexkatz/react-tiny-popover/blob/6db9f60cc6627987a4e57c971c924171ce16c0a1/src/Popover.tsx#L32)

I've looked at it a bit and I have tracked it down to the [call to `renderPopover`](https://github.com/alexkatz/react-tiny-popover/blob/6db9f60cc6627987a4e57c971c924171ce16c0a1/src/Popover.tsx#L82) within the updatePopover call. However, I am not sure how to deal with it as this class based component with hundreds of lifetime hooks is a bit out of my scope.

@willrogers are there any similar popover packages we could try out which might avoid this issue? Or can we live with it for the time being? I could create an issue but not sure how actively it is being worked on.

**Comment by TimGuiteDiamond** [comment]
Some styling issues persist, perhaps changes similar to similar to #113 are required?

Also need to provide a default description for actions so that context menu on right-click does not appear blank to user.

**Comment by TimGuiteDiamond** [comment]
Good job on flattening the JSON though, this will work nicely with #114

**Pull request merged**